File: src/core/cargo.ts
// Cargo hold management and mission order execution.
// Ships with cargo capacity carry items between colonies following sequential mission steps.

import type { ColonyState, MissionStep, ShipEntry } from "../types";
import { state } from "./state";

// Flat-packed items (prefixed "flat-") weigh 500 kg each; resources weigh 1 kg per unit.
const FLAT_PACKED_WEIGHT_KG = 500;
const RESOURCE_WEIGHT_KG = 1;

// Loading/unloading rate: 1000 kg per sim-day
const CARGO_TRANSFER_RATE_KG_PER_DAY = 1000;

export function getCargoWeightKg(ship: ShipEntry): number {
	let weight = 0;
	for (const [itemId, qty] of Object.entries(ship.cargoHold)) {
		weight += qty * (itemId.startsWith("flat-") ? FLAT_PACKED_WEIGHT_KG : RESOURCE_WEIGHT_KG);
	}
	return weight;
}

export function getCargoCapacityKg(ship: ShipEntry): number {
	const design = ship.designId ? state.shipDesigns.get(ship.designId) : undefined;
	return design?.cargoCapacityKg ?? 0;
}

/** Weight per unit for a given item id. */
function itemWeightKg(itemId: string): number {
	return itemId.startsWith("flat-") ? FLAT_PACKED_WEIGHT_KG : RESOURCE_WEIGHT_KG;
}

/** How many units of an item can still fit in the cargo hold. */
function remainingCapacityUnits(ship: ShipEntry, itemId: string): number {
	const freeKg = getCargoCapacityKg(ship) - getCargoWeightKg(ship);
	if (freeKg <= 0) return 0;
	return Math.floor(freeKg / itemWeightKg(itemId));
}

/** Find the colony at the ship's current orbit location, if any. */
function getColonyAtShip(ship: ShipEntry): ColonyState | undefined {
	return state.colonies.get(ship.hostPlanetName);
}

/** Resolve item source in a colony stockpile (resources or flatPacked). */
function getColonyStock(colony: ColonyState, itemId: string): number {
	if (itemId.startsWith("flat-")) {
		return colony.stockpile.flatPacked[itemId] ?? 0;
	}
	return colony.stockpile.resources[itemId] ?? 0;
}

function setColonyStock(colony: ColonyState, itemId: string, value: number): void {
	if (itemId.startsWith("flat-")) {
		colony.stockpile.flatPacked[itemId] = value;
	} else {
		colony.stockpile.resources[itemId] = value;
	}
}

function advanceStep(ship: ShipEntry): void {
	ship.missionOrderIndex++;
	if (ship.missionOrderIndex >= ship.missionOrders.length) {
		// Past the end with no repeat — stay at end (idle)
		ship.missionOrderIndex = ship.missionOrders.length;
	}
}

/** Execute a load-cargo step: transfer items from colony to ship at a rate per day. */
function tickLoadCargo(ship: ShipEntry, step: MissionStep, simDt: number): void {
	if (ship.shipState !== "orbiting") return;
	const colony = getColonyAtShip(ship);
	if (!colony) return;

	const itemId = step.itemId;
	if (!itemId) {
		advanceStep(ship);
		return;
	}

	const weightPerUnit = itemWeightKg(itemId);
	const maxUnitsByRate = Math.floor((CARGO_TRANSFER_RATE_KG_PER_DAY * simDt) / weightPerUnit);
	if (maxUnitsByRate <= 0) return;

	const colonyAvailable = getColonyStock(colony, itemId);
	const capacityUnits = remainingCapacityUnits(ship, itemId);
	const wantedUnits =
		step.quantity != null && step.quantity > 0
			? Math.max(0, step.quantity - (ship.cargoHold[itemId] ?? 0))
			: colonyAvailable;

	const transferUnits = Math.min(maxUnitsByRate, colonyAvailable, capacityUnits, wantedUnits);

	if (transferUnits > 0) {
		ship.cargoHold[itemId] = (ship.cargoHold[itemId] ?? 0) + transferUnits;
		setColonyStock(colony, itemId, colonyAvailable - transferUnits);
	}

	// Advance when done: colony empty, cargo full, or desired quantity reached
	const doneLoading =
		colonyAvailable - transferUnits <= 0 ||
		capacityUnits - transferUnits <= 0 ||
		(step.quantity != null && step.quantity > 0 && (ship.cargoHold[itemId] ?? 0) >= step.quantity);
	if (doneLoading) advanceStep(ship);
}

/** Execute an unload-cargo step: transfer items from ship to colony at a rate per day. */
function tickUnloadCargo(ship: ShipEntry, step: MissionStep, simDt: number): void {
	if (ship.shipState !== "orbiting") return;
	const colony = getColonyAtShip(ship);
	if (!colony) return;

	const itemId = step.itemId;
	if (!itemId) {
		advanceStep(ship);
		return;
	}

	const weightPerUnit = itemWeightKg(itemId);
	const maxUnitsByRate = Math.floor((CARGO_TRANSFER_RATE_KG_PER_DAY * simDt) / weightPerUnit);
	if (maxUnitsByRate <= 0) return;

	const shipAvailable = ship.cargoHold[itemId] ?? 0;
	const wantedUnits =
		step.quantity != null && step.quantity > 0
			? Math.min(step.quantity, shipAvailable)
			: shipAvailable;

	const transferUnits = Math.min(maxUnitsByRate, wantedUnits);

	if (transferUnits > 0) {
		ship.cargoHold[itemId] = shipAvailable - transferUnits;
		if (ship.cargoHold[itemId] <= 0) delete ship.cargoHold[itemId];
		setColonyStock(colony, itemId, getColonyStock(colony, itemId) + transferUnits);
	}

	// Advance when ship has no more of this item or desired quantity transferred
	const remaining = ship.cargoHold[itemId] ?? 0;
	if (remaining <= 0 || wantedUnits - transferUnits <= 0) advanceStep(ship);
}

/**
 * Returns true if the current mission step is "transfer-to" and the ship needs to move.
 * The caller (main.ts) is responsible for actually initiating the transfer via the
 * rendering layer, since cargo.ts cannot import rendering code.
 */
export function getMissionTransferTarget(ship: ShipEntry): string | null {
	if (ship.missionOrders.length === 0) return null;
	if (ship.missionOrderIndex >= ship.missionOrders.length) return null;
	const step = ship.missionOrders[ship.missionOrderIndex];
	if (step.type !== "transfer-to" || !step.target) return null;
	if (step.target === ship.hostPlanetName) return null; // already there
	return step.target;
}

/** Advance past a transfer-to step (called after arrival or when already at target). */
export function advanceMissionTransferStep(ship: ShipEntry): void {
	if (ship.missionOrders.length === 0) return;
	if (ship.missionOrderIndex >= ship.missionOrders.length) return;
	const step = ship.missionOrders[ship.missionOrderIndex];
	if (step.type === "transfer-to") advanceStep(ship);
}

/** Returns true if ship has active mission orders to execute. */
export function hasMissionOrders(ship: ShipEntry): boolean {
	return ship.missionOrders.length > 0 && ship.missionOrderIndex < ship.missionOrders.length;
}

/**
 * Tick the current mission step for a ship.
 * Called each frame from the ship tick loop. Transfer-to steps are handled externally
 * (main.ts calls getMissionTransferTarget + initiateTransfer). This function handles
 * load/unload/repeat steps only.
 */
export function tickMissionOrders(ship: ShipEntry, simDt: number): void {
	if (!hasMissionOrders(ship)) return;

	const step = ship.missionOrders[ship.missionOrderIndex];
	switch (step.type) {
		case "load-cargo":
			tickLoadCargo(ship, step, simDt);
			break;
		case "unload-cargo":
			tickUnloadCargo(ship, step, simDt);
			break;
		case "transfer-to":
			// If already at target, advance. Transfer initiation handled by main.ts.
			if (step.target === ship.hostPlanetName) advanceStep(ship);
			break;
		case "repeat":
			ship.missionOrderIndex = 0;
			break;
	}
}

File: rust-backend/crates/drift-sim/src/cargo.rs
// Cargo hold management and mission order execution.
// Ported from src/core/cargo.ts

use crate::state::{MissionStep, ShipEntry, State};

const FLAT_PACKED_WEIGHT_KG: f64 = 500.0;
const RESOURCE_WEIGHT_KG: f64 = 1.0;
const CARGO_TRANSFER_RATE_KG_PER_DAY: f64 = 1000.0;

fn item_weight_kg(item_id: &str) -> f64 {
    if item_id.starts_with("flat-") {
        FLAT_PACKED_WEIGHT_KG
    } else {
        RESOURCE_WEIGHT_KG
    }
}

/// Total cargo mass in kg across all items in the hold.
pub fn get_cargo_weight_kg(ship: &ShipEntry) -> f64 {
    ship.cargo_hold
        .iter()
        .map(|(id, qty)| qty * item_weight_kg(id))
        .sum()
}

/// Cargo capacity in kg from the ship's design. Returns 0 if no design is set.
pub fn get_cargo_capacity_kg(ship: &ShipEntry, state: &State) -> f64 {
    let design_id = match &ship.design_id {
        Some(id) => id,
        None => return 0.0,
    };
    state
        .ship_designs
        .get(design_id)
        .map(|d| d.cargo_capacity_kg)
        .unwrap_or(0.0)
}

/// How many units of `item_id` can still fit given remaining capacity.
fn remaining_capacity_units(ship: &ShipEntry, item_id: &str, state: &State) -> f64 {
    let free_kg = get_cargo_capacity_kg(ship, state) - get_cargo_weight_kg(ship);
    if free_kg <= 0.0 {
        return 0.0;
    }
    (free_kg / item_weight_kg(item_id)).floor()
}

/// Get available stock for an item from the colony at the ship's current orbit.
fn get_colony_stock(state: &State, host_planet_name: &str, item_id: &str) -> f64 {
    let colony = match state.colonies.get(host_planet_name) {
        Some(c) => c,
        None => return 0.0,
    };
    if item_id.starts_with("flat-") {
        *colony.stockpile.flat_packed.get(item_id).unwrap_or(&0) as f64
    } else {
        *colony.stockpile.resources.get(item_id).unwrap_or(&0.0)
    }
}

fn advance_step(ship: &mut ShipEntry) {
    ship.mission_order_index += 1;
}

fn tick_load_cargo(ship: &mut ShipEntry, step: &MissionStep, sim_dt: f64, state: &mut State) {
    if ship.ship_state != "orbiting" {
        return;
    }

    let host = ship.host_planet_name.clone();
    if !state.colonies.contains_key(&host) {
        return;
    }

    let item_id = match &step.item_id {
        Some(id) => id.clone(),
        None => {
            advance_step(ship);
            return;
        }
    };

    let weight_per_unit = item_weight_kg(&item_id);
    let max_units_by_rate = ((CARGO_TRANSFER_RATE_KG_PER_DAY * sim_dt) / weight_per_unit).floor();
    if max_units_by_rate <= 0.0 {
        return;
    }

    let colony_available = get_colony_stock(state, &host, &item_id);
    let capacity_units = remaining_capacity_units(ship, &item_id, state);
    let current_in_hold = *ship.cargo_hold.get(&item_id).unwrap_or(&0.0);

    let wanted_units = match step.quantity {
        Some(q) if q > 0.0 => (q - current_in_hold).max(0.0),
        _ => colony_available,
    };

    let transfer_units = max_units_by_rate
        .min(colony_available)
        .min(capacity_units)
        .min(wanted_units);

    if transfer_units > 0.0 {
        let entry = ship.cargo_hold.entry(item_id.clone()).or_insert(0.0);
        *entry += transfer_units;

        let colony = state.colonies.get_mut(&host).unwrap();
        if item_id.starts_with("flat-") {
            let stock = colony
                .stockpile
                .flat_packed
                .entry(item_id.clone())
                .or_insert(0);
            *stock = stock.saturating_sub(transfer_units as u32);
        } else {
            let stock = colony
                .stockpile
                .resources
                .entry(item_id.clone())
                .or_insert(0.0);
            *stock -= transfer_units;
        }
    }

    let new_in_hold = *ship.cargo_hold.get(&item_id).unwrap_or(&0.0);
    let new_colony_available = colony_available - transfer_units;
    let done = new_colony_available <= 0.0
        || remaining_capacity_units(ship, &item_id, state) <= 0.0
        || matches!(step.quantity, Some(q) if q > 0.0 && new_in_hold >= q);

    if done {
        advance_step(ship);
    }
}

fn tick_unload_cargo(ship: &mut ShipEntry, step: &MissionStep, sim_dt: f64, state: &mut State) {
    if ship.ship_state != "orbiting" {
        return;
    }

    let host = ship.host_planet_name.clone();
    if !state.colonies.contains_key(&host) {
        return;
    }

    let item_id = match &step.item_id {
        Some(id) => id.clone(),
        None => {
            advance_step(ship);
            return;
        }
    };

    let weight_per_unit = item_weight_kg(&item_id);
    let max_units_by_rate = ((CARGO_TRANSFER_RATE_KG_PER_DAY * sim_dt) / weight_per_unit).floor();
    if max_units_by_rate <= 0.0 {
        return;
    }

    let ship_available = *ship.cargo_hold.get(&item_id).unwrap_or(&0.0);
    let wanted_units = match step.quantity {
        Some(q) if q > 0.0 => q.min(ship_available),
        _ => ship_available,
    };

    let transfer_units = max_units_by_rate.min(wanted_units);

    if transfer_units > 0.0 {
        let new_amount = ship_available - transfer_units;
        if new_amount <= 0.0 {
            ship.cargo_hold.remove(&item_id);
        } else {
            *ship.cargo_hold.entry(item_id.clone()).or_insert(0.0) = new_amount;
        }

        let colony = state.colonies.get_mut(&host).unwrap();
        if item_id.starts_with("flat-") {
            let stock = colony
                .stockpile
                .flat_packed
                .entry(item_id.clone())
                .or_insert(0);
            *stock += transfer_units as u32;
        } else {
            let stock = colony
                .stockpile
                .resources
                .entry(item_id.clone())
                .or_insert(0.0);
            *stock += transfer_units;
        }
    }

    let remaining = *ship.cargo_hold.get(&item_id).unwrap_or(&0.0);
    let done = remaining <= 0.0 || wanted_units - transfer_units <= 0.0;
    if done {
        advance_step(ship);
    }
}

/// Returns true if the ship has active mission orders to execute.
pub fn has_mission_orders(ship: &ShipEntry) -> bool {
    !ship.mission_orders.is_empty() && ship.mission_order_index < ship.mission_orders.len()
}

/// Returns the transfer target if the current step is a pending "transfer-to".
/// Returns None if already at the target, no orders, or wrong step type.
pub fn get_mission_transfer_target(ship: &ShipEntry) -> Option<String> {
    if ship.mission_orders.is_empty() {
        return None;
    }
    let step = ship.mission_orders.get(ship.mission_order_index)?;
    if step.step_type != "transfer-to" {
        return None;
    }
    let target = step.target.as_deref()?;
    if target == ship.host_planet_name {
        return None;
    }
    Some(target.to_string())
}

/// Advance past a "transfer-to" step (called after arrival).
pub fn advance_mission_transfer_step(ship: &mut ShipEntry) {
    if ship.mission_orders.is_empty() {
        return;
    }
    if ship.mission_order_index >= ship.mission_orders.len() {
        return;
    }
    if ship.mission_orders[ship.mission_order_index].step_type == "transfer-to" {
        advance_step(ship);
    }
}

/// Tick the current mission step. Handles load/unload/repeat/transfer-to steps.
pub fn tick_mission_orders(ship: &mut ShipEntry, sim_dt: f64, state: &mut State) {
    if !has_mission_orders(ship) {
        return;
    }

    let step = ship.mission_orders[ship.mission_order_index].clone();
    match step.step_type.as_str() {
        "load-cargo" => tick_load_cargo(ship, &step, sim_dt, state),
        "unload-cargo" => tick_unload_cargo(ship, &step, sim_dt, state),
        "transfer-to" => {
            if step.target.as_deref() == Some(ship.host_planet_name.as_str()) {
                advance_step(ship);
            }
        }
        "repeat" => {
            ship.mission_order_index = 0;
        }
        _ => {}
    }
}
